staging/src/k8s.io/apiserver/pkg/util/proxy/websocket.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/httpstream/wsstream"
+	constants "k8s.io/apimachinery/pkg/util/remotecommand"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// idleTimeout is the read/write deadline set for websocket server connection. Reading
+	// or writing the connection will return an i/o timeout if this deadline is exceeded.
+	// Currently, we use the same value as the kubelet websocket server.
+	defaultIdleConnectionTimeout = 4 * time.Hour
+
+	// Deadline for writing errors to the websocket connection before io/timeout.
+	writeErrorDeadline = 10 * time.Second
+)
+
+// Options contains details about which streams are required for
+// remote command execution.
+type Options struct {
+	Stdin  bool
+	Stdout bool
+	Stderr bool
+	Tty    bool
+}
+
+// conns contains the connection and streams used when
+// forwarding an attach or execute session into a container.
+type conns struct {
+	conn         io.Closer
+	stdinStream  io.ReadCloser
+	stdoutStream io.WriteCloser
+	stderrStream io.WriteCloser
+	writeStatus  func(status *apierrors.StatusError) error
+	resizeStream io.ReadCloser
+	resizeChan   chan remotecommand.TerminalSize
+	tty          bool
+}
+
+// Create WebSocket server streams to respond to a WebSocket client. Creates the streams passed
+// in the stream options.
+func webSocketServerStreams(req *http.Request, w http.ResponseWriter, opts Options) (*conns, error) {
+	ctx, err := createWebSocketStreams(req, w, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if ctx.resizeStream != nil {
+		ctx.resizeChan = make(chan remotecommand.TerminalSize)
+		go func() {
+			// Ensure any panic in this child goroutine does not crash server. Log panic
+			// error since it it not possible to propagate panic to parent goroutine.
+			defer func() {
+				if p := recover(); p != nil {
+					klog.Errorf("websocket server panic while handling terminal resizes")
```

Comment:
HandleCrash already logs, we don't need to log again, add a comment that the resize channel closes in panic case and does not take down the caller

Reply:
Removed `klog` and added standard `runtime` panic logging instead.